### PR TITLE
Output check/record: fixes and output support

### PR DESCRIPTION
**Pull request by clebergnu**
The last work on the output check/record feature introduced the `avocado.utils.process` support for generating an combined output file.

This PR builds on that and provides support for checking against a combined output file, does a lot of refactors, adds many tests, fixes bugs and updates the documentation.

There's one last feature gap that I'll post later: the ability to configure the reference file names (`stdout.expected`, `stderr.expected` and `output.expected).

### CODE REVIEW

File: selftests/functional/test_output_check.py [hunk 1]
```diff
@@ -30,10 +30,10 @@ def setUp(self):
             'avocado_output_check_functional')
         self.output_script.save()
 
-    def test_output_record_none(self):
+    def _check_output_record_all(self):
```
Comment (apahim) [inline selftests/functional/test_output_check.py]: ok

File: avocado/core/test.py [hunk 2]
```diff
@@ -698,23 +698,33 @@ def _stop_logging(self):
...
+
+        :param produced_file_path: the location of the file that was produced
+                                   by this test execution
+        :type produced_file_path: str
+        :param reference_file_name: the name of the file that will be used on
+                                    subsequent runs to check the test produced
+                                    the correct content.  This file will be
+                                    saved into a location obtained by
+                                    calling :meth:`get_data()`.
+        :type reference_file_name: str
```
Comment (apahim) [inline avocado/core/test.py]: Ok. Very nice docstring.

File: avocado/core/test.py [hunk 3]
```diff
@@ -726,61 +726,54 @@ def _record_reference(self, produced_file_path, reference_file_name):
...
+        :type child_log_name: str
+        :param name: optional parameter for a descriptive name of the type of
+                     content being checked here
+        :type name: str
+        '''
+        reference_path = self.get_data(reference_file_name)
+        if reference_path is not None:
+            expected = genio.read_file(reference_path)
+            actual = genio.read_file(produced_file_path)
+            diff_path = os.path.join(self.logdir, diff_file_name)
```
Comment (apahim) [inline avocado/core/test.py]: Ok, another nice refactor with even nicer docstring.

File: selftests/unit/test_test.py [hunk 4]
```diff
@@ -127,6 +127,42 @@ def test_try_override_test_variable(self):
...
+        Tests that a check is not made for a file that does not exist
+        '''
+        tst = self.DummyTest("test", test.TestID(1, "test"),
+                             base_logdir=self.tmpdir)
+        self.assertFalse(tst._check_reference('does_not_exist',
+                                              'stdout.expected',
+                                              'stdout.diff',
+                                              'stdout_diff',
+                                              'Stdout'))
+
```
Comment (apahim) [inline selftests/unit/test_test.py]: ok

File: avocado/core/test.py [hunk 5]
```diff
@@ -877,26 +887,41 @@ def _run_avocado(self):
...
+                        try:
+                            self._check_reference(self._stderr_file,
+                                                  'stderr.expected',
+                                                  'stderr.diff',
+                                                  'stderr_diff',
+                                                  'Stderr')
+                        except Exception as details:
+                            stacktrace.log_exc_info(sys.exc_info(),
+                                                    logger=LOG_JOB)
+                            stderr_check_exception = details
```
Comment (apahim) [inline avocado/core/test.py]: Yep, that was the missing piece of this new 'output.expected' feature.

Comment (apahim) [reply avocado/core/test.py]: Please add a selftest for it.

File: avocado/utils/script.py [hunk 6]
```diff
@@ -29,6 +29,9 @@
                 stat.S_IRGRP | stat.S_IWGRP | stat.S_IXGRP |
                 stat.S_IROTH | stat.S_IXOTH)
 
+#: What is commonly known as "0444" or "u=r,g=r,o=r"
+READ_ONLY_MODE = (stat.S_IRUSR | stat.S_IRGRP | stat.S_IROTH)
+
```
Comment (apahim) [inline avocado/utils/script.py]: ok

File: avocado/core/test.py [hunk 7]
```diff
@@ -879,60 +879,58 @@ def _run_avocado(self):
...
+                    self._record_reference(self._stdout_file,
+                                           "stdout.expected")
+                if output_check_record in ['all', 'both', 'stderr']:
+                    self._record_reference(self._stderr_file,
+                                           "stderr.expected")
+
+            # check the output and produce test failures
+            if ((not job_standalone or
+                 output_check_record != 'none') and output_check == 'on'):
+                output_checked = False
```
Comment (apahim) [inline avocado/core/test.py]: yes, that make sthe check to happen when it's actually required.

File: avocado/core/test.py [hunk 8]
```diff
@@ -879,60 +879,58 @@ def _run_avocado(self):
...
+                    try:
+                        self._check_reference(self._stderr_file,
+                                              'stderr.expected',
+                                              'stderr.diff',
+                                              'stderr_diff',
+                                              'Stderr')
+                    except Exception as details:
+                        stacktrace.log_exc_info(sys.exc_info(),
+                                                logger=LOG_JOB)
+                        stderr_check_exception = details
```
Comment (apahim) [inline avocado/core/test.py]: ok

File: avocado/utils/process.py [hunk 9]
```diff
@@ -542,13 +548,19 @@ def _init_subprocess(self):
                 self._combined_drainer.start()
 
             else:
+                if self.allow_output_check == 'none':
+                    stdout_stream_logger = None
+                    stderr_stream_logger = None
+                else:
+                    stdout_stream_logger = stdout_log
+                    stderr_stream_logger = stderr_log
```
Comment (apahim) [inline avocado/utils/process.py]: ok, another good one.

File: selftests/functional/test_output.py [hunk 10]
```diff
@@ -213,6 +237,33 @@ def test_check_record_no_module_default(self):
...
+            self.assertTrue(os.path.exists(stdout_path))
+            with open(stdout_path, 'r') as stdout:
+                self.assertEqual(stdout.read(),
+                                 '__STDOUT_CONTENT____STDOUT_DO_RECORD_CONTENT__')
+            stderr_path = os.path.join(testdir, 'stderr')
+            self.assertTrue(os.path.exists(stderr_path))
+            with open(stderr_path, 'r') as stderr:
+                self.assertEqual(stderr.read(),
+                                 '__STDERR_CONTENT____STDERR_DO_RECORD_CONTENT__')
+
```
Comment (apahim) [inline selftests/functional/test_output.py]: very nice test

File: docs/source/WritingTests.rst [hunk 11]
```diff
@@ -752,52 +752,88 @@ Test Output Check and Output Record Mode
...
+
+You can disable this automatic check when a reference file exists by
+passing ``--output-check=off`` to the test runner.
+
+.. tip:: The :mod:`avocado.utils.process` APIs have a parameter called
+         ``allow_output_check`` that let you individually select the
+         output that will be part of the test output and recorded
+         reference files.  Some other APIs built on top of
+         :mod:`avocado.utils.process`, such as the ones in
+         :mod:`avocado.utils.build` also provide the same parameter.
```
Comment (apahim) [inline docs/source/WritingTests.rst]: Ok, very nice documentation. The *tip* is a nice bonus.

**Comment by clebergnu** [comment]
Closing in favor of #2317

**Pull request closed without merging**